src/component_selector.jl

```diff
@@ -0,0 +1,434 @@
+   zero components, else return an iterator of the one component
+ - there is a sensible default for `get_groups`; it MAY be overridden
+=#
+"`ComponentSelector`s that can only refer to zero or one components."
+abstract type SingularComponentSelector <: ComponentSelector end
+
+#=
+`PluralComponentSelector` extension notes:
+The interface is the same as for `ComponentSelector`.
+=#
+"""`ComponentSelector`s that may refer to multiple components."""
+abstract type PluralComponentSelector <: ComponentSelector end
+
+#=
+`DynamicallyGroupedComponentSelector` extension notes:
+One MAY subtype this and have a `groupby::Union{Symbol, Function}` field to get an automatic
+implementation of `get_groups`.
+=#
+"`PluralComponentSelector`s whose grouping is determined by a `groupby` field."
+abstract type DynamicallyGroupedComponentSelector <: PluralComponentSelector end
+
+# COMMON COMPONENTSELECTOR INFRASTRUCTURE
+"Canonical way to turn an `InfrastructureSystemsComponent` subtype into a unique string."
+subtype_to_string(subtype::Type{<:InfrastructureSystemsComponent}) =
+    strip_module_name(subtype)
+
+"""
+Canonical way to turn an `InfrastructureSystemsComponent` specification/instance into a
+unique-per-system string.
+"""
+component_to_qualified_string(
+    component_type::Type{<:InfrastructureSystemsComponent},
+    component_name::AbstractString,
+) = subtype_to_string(component_type) * COMPONENT_NAME_DELIMITER * component_name
+component_to_qualified_string(component::InfrastructureSystemsComponent) =
+    component_to_qualified_string(typeof(component), get_name(component))
+
+const VALID_GROUPBY_KEYWORDS = [:all, :each]
+const DEFAULT_GROUPBY = :each
+
+"""
+Helper function to check that the `groupby` argument is valid. Passes it through if so,
+errors if not.
+"""
+validate_groupby(groupby::Symbol) =
+    if (groupby in VALID_GROUPBY_KEYWORDS)
+        groupby
+    else
+        throw(ArgumentError("groupby must be one of $VALID_GROUPBY_KEYWORDS or a function"))
+    end
+validate_groupby(groupby::Function) = groupby  # Don't try to validate functions for now
+
+# Generic implementations/generic docstrings for simple functions with many methods
+"""
+Factory function to create the appropriate subtype of `ComponentSelector` given the
+arguments. Users should call this rather than manually constructing `ComponentSelector`
+subtypes.
+"""
+function make_selector end
+
+# Override this if you define a ComponentSelector subtype with no name field
+"""
+Get the name of the `ComponentSelector`. This is either the default name or a custom name
+passed in at creation time.
+"""
+get_name(selector::ComponentSelector) = selector.name
+
+"""
+    get_available_components(selector, sys)
+Get the available components of the collection that make up the `ComponentSelector`.
+"""
+get_available_components(selector::ComponentSelector, sys) =
+    get_components(selector, sys; scope_limiter = x -> get_available(sys, x))
```

Comment:
I'm a bit concerned that people will use this as example code. And then eventually ask why scope_limiter isn't a kwarg in the System get_components method. Wouldn't it be better to follow the Julia convention and pass the "scope_limiter" as the first argument?

Reply:
The goals with this design are to (a) keep `scope_limiter` out of the public interface, as we resolved earlier, while (b) keeping it in a sort of "developer interface" so that it can be used for `get_available_components`. Absent another layer of indirection (have each `ComponentSelector` subtype implement `_get_components` with a `scope_limiter` and have `get_components` and `get_available_components` call that), putting `scope_limiter` as the first argument would more or less _force_ it be part of the public interface: it would explicitly appear in method lists and such. Do you think it's worth it to add that layer of indirection, or is the current solution (perhaps with a comment added here that users are not to take this source code as example code?) good enough?